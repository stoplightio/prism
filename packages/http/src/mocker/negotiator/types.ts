import { IHttpOperationConfig } from '@stoplight/prism-http';
import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}

export type NegotiationOptions = IHttpOperationConfig;

export type NegotiatePartialOptions = {
  code: string;
  dynamic: boolean;
  exampleKey?: string;
};
