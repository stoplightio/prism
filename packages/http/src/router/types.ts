import { IHttpOperationEx } from '../types';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperationEx;
  pathMatch: MatchType;
  methodMatch: MatchType;
  serverMatch?: Nullable<MatchType>;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
