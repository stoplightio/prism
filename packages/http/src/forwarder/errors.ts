import type { ProblemJson } from '..';

export const UPSTREAM_NOT_IMPLEMENTED: Omit<ProblemJson, 'detail'> = {
  type: 'UPSTREAM_NOT_IMPLEMENTED',
  title: 'The server does not support the functionality required to fulfill the request',
  status: 501,
};

export const PROXY_UNSUPPORTED_REQUEST_BODY: Omit<ProblemJson, 'detail'> = {
  type: 'PROXY_UNSUPPORTED_REQUEST_BODY',
  title:
    'The Prism proxy does not support sending a GET/HEAD request with a message body to an upstream server. See: https://github.com/stoplightio/prism/issues/2259',
  status: 501,
};
