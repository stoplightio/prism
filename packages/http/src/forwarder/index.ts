import axios from 'axios';
import { HttpForwarder } from './HttpForwarder';

const forwarder = new HttpForwarder();

export { forwarder };
export const CancelToken = axios.CancelToken;
