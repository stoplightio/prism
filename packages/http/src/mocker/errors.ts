import { ProblemJson } from '../types';

export const UNPROCESSABLE_ENTITY: Omit<ProblemJson, 'detail'> = {
  type: 'UNPROCESSABLE_ENTITY',
  title: 'Invalid request body payload',
  status: 422,
};

export const NOT_ACCEPTABLE: Omit<ProblemJson, 'detail'> = {
  type: 'NOT_ACCEPTABLE',
  title: 'The server cannot produce a representation for your accept header',
  status: 406,
};

export const NOT_FOUND: Omit<ProblemJson, 'detail'> = {
  type: 'NOT_FOUND',
  title: 'The server cannot find the requested content',
  status: 404,
};

export const UNAUTHORIZED: Omit<ProblemJson, 'detail'> = {
  type: 'UNAUTHORIZED',
  title: 'Invalid security scheme used',
  status: 401,
};

export const VIOLATIONS: Omit<ProblemJson, 'detail'> = {
  type: 'VIOLATIONS',
  title: 'Request/Response not valid',
  status: 500,
};
