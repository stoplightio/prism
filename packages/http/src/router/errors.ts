import { ProblemJson } from '../types';

export const NO_BASE_URL_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Attempted to make a request to a server but neither baseUrl param provided nor servers defined in the spec',
  type: 'NO_BASE_URL_ERROR',
  status: 400,
};
export const NO_RESOURCE_PROVIDED_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Route not resolved, no resource provided',
  type: 'NO_RESOURCE_PROVIDED_ERROR',
  status: 404,
};
export const NO_PATH_MATCHED_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Route not resolved, no path matched',
  type: 'NO_PATH_MATCHED_ERROR',
  status: 404,
};
export const NO_SERVER_MATCHED_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Route not resolved, no server matched',
  type: 'NO_SERVER_MATCHED_ERROR',
  status: 404,
};
export const NO_METHOD_MATCHED_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Route resolved, but no method matched',
  type: 'NO_METHOD_MATCHED_ERROR',
  status: 405,
};
export const NO_SERVER_CONFIGURATION_PROVIDED_ERROR: Omit<ProblemJson, 'detail'> = {
  title: 'Route not resolved, no server configuration provided',
  type: 'NO_SERVER_CONFIGURATION_PROVIDED_ERROR',
  status: 404,
};
