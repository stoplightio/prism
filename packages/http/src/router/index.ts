import { IPrismComponents } from '@stoplight/prism-core';
import { IHttpOperation, IServer } from '@stoplight/types';
import * as E from 'fp-ts/lib/Either';
import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/pipeable';
import { IHttpConfig, IHttpRequest, ProblemJsonError } from '../types';
import {
  NO_METHOD_MATCHED_ERROR,
  NO_PATH_MATCHED_ERROR,
  NO_RESOURCE_PROVIDED_ERROR,
  NO_SERVER_CONFIGURATION_PROVIDED_ERROR,
  NO_SERVER_MATCHED_ERROR,
} from './errors';
import { matchBaseUrl } from './matchBaseUrl';
import { matchPath } from './matchPath';
import { IMatch, MatchType } from './types';

const eitherSequence = A.array.sequence(E.either);

const route: IPrismComponents<IHttpOperation, IHttpRequest, unknown, IHttpConfig>['route'] = ({ resources, input }) => {
  const { path: requestPath, baseUrl: requestBaseUrl } = input.url;

  if (!requestPath.startsWith('/')) {
    return E.left(new Error(`The request path '${requestPath}' must start with a slash.`));
  }

  return pipe(
    resources,
    E.fromPredicate(A.isNonEmpty, () =>
      ProblemJsonError.fromTemplate(
        NO_RESOURCE_PROVIDED_ERROR,
        `The current document does not have any resource to match with.`
      )
    ),
    E.chain(resources =>
      eitherSequence(
        resources.map(resource =>
          pipe(
            matchPath(requestPath, resource.path),
            E.chain<
              Error,
              MatchType,
              { pathMatch: MatchType; methodMatch: MatchType; serverMatch?: MatchType; resource: IHttpOperation }
            >(pathMatch => {
              if (pathMatch === MatchType.NOMATCH)
                return E.right({
                  pathMatch,
                  methodMatch: MatchType.NOMATCH,
                  resource,
                });

              const methodMatch = matchByMethod(input, resource) ? MatchType.CONCRETE : MatchType.NOMATCH;

              if (methodMatch === MatchType.NOMATCH) {
                return E.right({
                  pathMatch,
                  methodMatch,
                  resource,
                });
              }

              const { servers = [] } = resource;

              if (requestBaseUrl && servers.length > 0) {
                const serverMatchEither = matchServer(servers, requestBaseUrl);
                return pipe(
                  serverMatchEither,
                  E.map(serverMatch => ({ pathMatch, methodMatch, serverMatch, resource }))
                );
              }

              return E.right({
                pathMatch,
                methodMatch,
                resource,
              });
            })
          )
        )
      )
    ),
    E.chain(candidateMatches => {
      const pathMatches = candidateMatches.filter(match => match.pathMatch !== MatchType.NOMATCH);

      if (!pathMatches.length) {
        return E.left(
          ProblemJsonError.fromTemplate(
            NO_PATH_MATCHED_ERROR,
            `The route ${requestPath} hasn't been found in the specification file`
          )
        );
      }

      const methodMatches = pathMatches.filter(match => match.methodMatch !== MatchType.NOMATCH);

      if (!methodMatches.length) {
        return E.left(
          ProblemJsonError.fromTemplate(
            NO_METHOD_MATCHED_ERROR,
            `The route ${requestPath} has been matched, but it does not have "${input.method}" method defined`
          )
        );
      }

      if (requestBaseUrl) {
        if (resources.every(resource => !resource.servers || resource.servers.length === 0)) {
          return E.left(
            ProblemJsonError.fromTemplate(
              NO_SERVER_CONFIGURATION_PROVIDED_ERROR,
              `No server configuration has been provided, although ${requestBaseUrl} is set as server url`
            )
          );
        }

        const serverMatches = methodMatches.filter(
          match => !!match.serverMatch && match.serverMatch !== MatchType.NOMATCH
        );

        if (!serverMatches.length) {
          return E.left(
            ProblemJsonError.fromTemplate(
              NO_SERVER_MATCHED_ERROR,
              `The server url ${requestBaseUrl} hasn't been matched with any of the provided servers`
            )
          );
        }

        return E.right(disambiguateMatches(serverMatches));
      }

      return E.right(disambiguateMatches(methodMatches));
    })
  );
};

function matchServer(servers: IServer[], requestBaseUrl: string): E.Either<Error, MatchType> {
  return pipe(
    servers.map(server => matchBaseUrl(server, requestBaseUrl)),
    eitherSequence,
    E.map(matches => matches.filter(match => match !== MatchType.NOMATCH)),
    E.map(disambiguateServers)
  );
}

function matchByMethod(request: IHttpRequest, operation: IHttpOperation): boolean {
  return operation.method.toLowerCase() === request.method.toLowerCase();
}

function disambiguateMatches(matches: IMatch[]): IHttpOperation {
  const matchResult =
    // prefer concrete server and concrete path
    matches.find(match => areServerAndPath(match, MatchType.CONCRETE, MatchType.CONCRETE)) ||
    // then prefer templated server and concrete path
    matches.find(match => areServerAndPath(match, MatchType.TEMPLATED, MatchType.CONCRETE)) ||
    // then prefer concrete server and templated path
    matches.find(match => areServerAndPath(match, MatchType.CONCRETE, MatchType.TEMPLATED)) ||
    // then fallback to first
    matches[0];

  return matchResult.resource;
}

function areServerAndPath(match: IMatch, serverType: MatchType, pathType: MatchType) {
  const serverMatch = match.serverMatch;
  if (!serverMatch) {
    // server match will only be null if server matching is disabled.
    // therefore skip comparison.
    return match.pathMatch === pathType;
  }
  return serverMatch === serverType && match.pathMatch === pathType;
}

/**
 * If a concrete server match exists then return first such match.
 * If no concrete server match exists return first (templated) match.
 */
function disambiguateServers(serverMatches: MatchType[]): MatchType {
  const concreteMatch = serverMatches.find(serverMatch => serverMatch === MatchType.CONCRETE);
  return concreteMatch || serverMatches[0] || MatchType.NOMATCH;
}

export default route;
