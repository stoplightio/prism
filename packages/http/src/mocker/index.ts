import { IPrismComponents, IPrismInput } from '@stoplight/prism-core';
import {
  DiagnosticSeverity,
  IHttpHeaderParam,
  IHttpOperation,
  INodeExample,
  IMediaTypeContent,
} from '@stoplight/types';

import * as caseless from 'caseless';
import * as E from 'fp-ts/lib/Either';
import * as Record from 'fp-ts/lib/Record';
import { pipe } from 'fp-ts/lib/pipeable';
import * as R from 'fp-ts/lib/Reader';
import * as O from 'fp-ts/lib/Option';
import * as RE from 'fp-ts/lib/ReaderEither';
import { map } from 'fp-ts/lib/Array';
import { isNumber, isString, keyBy, mapValues, groupBy, get } from 'lodash';
import { Logger } from 'pino';
import * as typeIs from 'type-is';
import {
  ContentExample,
  IHttpOperationConfig,
  IHttpRequest,
  IHttpResponse,
  IMockHttpConfig,
  PayloadGenerator,
  ProblemJsonError,
} from '../types';
import withLogger from '../withLogger';
import { UNAUTHORIZED, UNPROCESSABLE_ENTITY } from './errors';
import { generate, generateStatic } from './generator/JSONSchema';
import helpers from './negotiator/NegotiatorHelpers';
import { IHttpNegotiationResult } from './negotiator/types';
import { runCallback } from './callback/callbacks';
import {
  decodeUriEntities,
  deserializeFormBody,
  findContentByMediaTypeOrFirst,
  splitUriParams,
} from '../validator/validators/body';
import { sequenceT } from 'fp-ts/lib/Apply';

const eitherRecordSequence = Record.sequence(E.either);
const eitherSequence = sequenceT(E.either);

const mock: IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IMockHttpConfig>['mock'] = ({
  resource,
  input,
  config,
}) => {
  const payloadGenerator: PayloadGenerator = config.dynamic ? generate : generateStatic;

  return pipe(
    withLogger(logger => {
      // setting default values
      const acceptMediaType = input.data.headers && caseless(input.data.headers).get('accept');
      if (!config.mediaTypes && acceptMediaType) {
        logger.info(`Request contains an accept header: ${acceptMediaType}`);
        config.mediaTypes = acceptMediaType.split(',');
      }

      return config;
    }),
    R.chain(mockConfig => negotiateResponse(mockConfig, input, resource)),
    R.chain(result => assembleResponse(result, payloadGenerator)),
    R.chain(response =>
      /*  Note: This is now just logging the errors without propagating them back. This might be moved as a first
          level concept in Prism.
      */
      logger =>
        pipe(
          response,
          E.map(response => runCallbacks({ resource, request: input.data, response })(logger)),
          E.chain(() => response)
        )
    )
  );
};

function runCallbacks({
  resource,
  request,
  response,
}: {
  resource: IHttpOperation;
  request: IHttpRequest;
  response: IHttpResponse;
}) {
  return withLogger(logger =>
    pipe(
      O.fromNullable(resource.callbacks),
      O.map(callbacks =>
        pipe(
          callbacks,
          map(callback =>
            runCallback({ callback, request: parseBodyIfUrlEncoded(request, resource), response })(logger)()
          )
        )
      )
    )
  );
}

/*
  This function should not be here at all, but unfortunately due to some limitations of the Monad we're using (Either)
  we cannot carry parsed informations in case of an error â€” which is what we do need instead.
*/
function parseBodyIfUrlEncoded(request: IHttpRequest, resource: IHttpOperation) {
  const mediaType = caseless(request.headers || {}).get('content-type');
  if (!mediaType) return request;

  if (!typeIs.is(mediaType, ['application/x-www-form-urlencoded'])) return request;

  const specs = pipe(
    O.fromNullable(resource.request),
    O.mapNullable(request => request.body),
    O.mapNullable(body => body.contents),
    O.getOrElse(() => [] as IMediaTypeContent[])
  );

  const encodedUriParams = splitUriParams(request.body as string);

  if (specs.length < 1) {
    return Object.assign(request, { body: encodedUriParams });
  }

  const content = pipe(
    O.fromNullable(mediaType),
    O.chain(mediaType => findContentByMediaTypeOrFirst(specs, mediaType)),
    O.map(({ content }) => content),
    O.getOrElse(() => specs[0] || {})
  );

  const encodings = get(content, 'encodings', []);

  if (!content.schema) return Object.assign(request, { body: encodedUriParams });

  return Object.assign(request, {
    body: deserializeFormBody(content.schema, encodings, decodeUriEntities(encodedUriParams)),
  });
}

function handleInputValidation(input: IPrismInput<IHttpRequest>, resource: IHttpOperation) {
  const securityValidation = input.validations.find(validation => validation.code === 401);

  return pipe(
    withLogger(logger => logger.warn({ name: 'VALIDATOR' }, 'Request did not pass the validation rules')),
    R.chain(() =>
      pipe(
        helpers.negotiateOptionsForInvalidRequest(resource.responses, securityValidation ? ['401'] : ['422', '400']),
        RE.mapLeft(() =>
          securityValidation
            ? ProblemJsonError.fromTemplate(
                UNAUTHORIZED,
                '',
                securityValidation.tags && securityValidation.tags.length
                  ? {
                      headers: { 'WWW-Authenticate': securityValidation.tags.join(',') },
                    }
                  : undefined
              )
            : ProblemJsonError.fromTemplate(
                UNPROCESSABLE_ENTITY,
                'Your request is not valid and no HTTP validation response was found in the spec, so Prism is generating this error for you.',
                {
                  validation: input.validations.map(detail => ({
                    location: detail.path,
                    severity: DiagnosticSeverity[detail.severity],
                    code: detail.code,
                    message: detail.message,
                  })),
                }
              )
        )
      )
    )
  );
}

function negotiateResponse(
  mockConfig: IHttpOperationConfig,
  input: IPrismInput<IHttpRequest>,
  resource: IHttpOperation
) {
  const { [DiagnosticSeverity.Error]: errors, [DiagnosticSeverity.Warning]: warnings } = groupBy(
    input.validations,
    validation => validation.severity
  );

  if (errors) {
    return handleInputValidation(input, resource);
  } else {
    return pipe(
      withLogger(logger => {
        warnings && warnings.forEach(warn => logger.warn({ name: 'VALIDATOR' }, warn.message));
        return logger.success(
          { name: 'VALIDATOR' },
          'The request passed the validation rules. Looking for the best response'
        );
      }),
      R.chain(() => helpers.negotiateOptionsForValidRequest(resource, mockConfig))
    );
  }
}

function assembleResponse(
  result: E.Either<Error, IHttpNegotiationResult>,
  payloadGenerator: PayloadGenerator
): R.Reader<Logger, E.Either<Error, IHttpResponse>> {
  return logger =>
    pipe(
      result,
      E.chain(negotiationResult =>
        pipe(
          eitherSequence(
            computeBody(negotiationResult, payloadGenerator),
            computeMockedHeaders(negotiationResult.headers || [], payloadGenerator)
          ),
          E.map(([mockedBody, mockedHeaders]) => {
            const response: IHttpResponse = {
              statusCode: parseInt(negotiationResult.code),
              headers: {
                ...mockedHeaders,
                ...(negotiationResult.mediaType && { 'Content-type': negotiationResult.mediaType }),
              },
              body: mockedBody,
            };

            logger.success(`Responding with the requested status code ${response.statusCode}`);

            return response;
          })
        )
      )
    );
}

function isINodeExample(nodeExample: ContentExample | undefined): nodeExample is INodeExample {
  return !!nodeExample && 'value' in nodeExample;
}

function computeMockedHeaders(headers: IHttpHeaderParam[], payloadGenerator: PayloadGenerator) {
  return eitherRecordSequence(
    mapValues(
      keyBy(headers, h => h.name),
      header => {
        if (header.schema) {
          if (header.examples && header.examples.length > 0) {
            const example = header.examples[0];
            if (isINodeExample(example)) {
              return E.right(example.value);
            }
          } else {
            return pipe(
              payloadGenerator(header.schema),
              E.map(example => {
                if (isNumber(example) || isString(example)) return example;
                return null;
              })
            );
          }
        }
        return E.right(null);
      }
    )
  );
}

function computeBody(
  negotiationResult: Pick<IHttpNegotiationResult, 'schema' | 'mediaType' | 'bodyExample'>,
  payloadGenerator: PayloadGenerator
): E.Either<Error, unknown> {
  if (isINodeExample(negotiationResult.bodyExample) && negotiationResult.bodyExample.value !== undefined) {
    return E.right(negotiationResult.bodyExample.value);
  } else if (negotiationResult.schema) {
    return payloadGenerator(negotiationResult.schema);
  }
  return E.right(undefined);
}

export default mock;
