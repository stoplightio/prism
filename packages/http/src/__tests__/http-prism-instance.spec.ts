import { IPrism } from '@stoplight/prism-core';
import { IHttpOperation } from '@stoplight/types/http-spec';
import { relative, resolve } from 'path';
import { createInstance, IHttpConfig, IHttpRequest, IHttpResponse, ProblemJsonError } from '../';
import { forwarder } from '../forwarder';
import { UNPROCESSABLE_ENTITY } from '../mocker/errors';
import { NO_PATH_MATCHED_ERROR } from '../router/errors';

describe('Http Prism Instance function tests', () => {
  let prism: IPrism<IHttpOperation, IHttpRequest, IHttpResponse, IHttpConfig, { path: string }>;

  beforeAll(async () => {
    prism = createInstance({ mock: { dynamic: false } }, {});
    await prism.load({
      path: relative(process.cwd(), resolve(__dirname, 'fixtures', 'no-refs-petstore-minimal.oas2.json')),
    });
  });

  test('keeps the instances separate', async () => {
    const second_prism = createInstance({ mock: { dynamic: false } }, {});
    await second_prism.load({
      path: relative(process.cwd(), resolve(__dirname, 'fixtures', 'no-refs-petstore-minimal.oas2.json')),
    });

    expect(prism.resources).toStrictEqual(second_prism.resources);
  });

  test('given incorrect route should throw error', () => {
    return expect(
      prism.process({
        method: 'get',
        url: {
          path: '/invalid-route',
        },
      }),
    ).rejects.toThrowError(ProblemJsonError.fromTemplate(NO_PATH_MATCHED_ERROR));
  });

  test('given correct route should return correct response', async () => {
    const response = await prism.process({
      method: 'get',
      url: {
        path: '/pet/findByStatus',
        query: {
          status: ['available', 'pending'],
        },
      },
    });

    const parsedBody = response!.output!.body;

    expect(parsedBody.length).toBeGreaterThan(0);

    parsedBody.forEach((element: any) => {
      expect(typeof element.name).toBe('string');
      expect(element.photoUrls).toBeInstanceOf(Array);
      expect(element.photoUrls.length).toBeGreaterThan(0);
    });

    expect(response).toMatchSnapshot({
      output: {
        body: expect.anything(),
      },
    });
  });

  test('given route with invalid param should return a validation error', () => {
    return expect(
      prism.process({
        method: 'get',
        url: {
          path: '/pet/findByStatus',
        },
      }),
    ).rejects.toThrowError(ProblemJsonError.fromTemplate(UNPROCESSABLE_ENTITY));
  });

  test('should support collection format multi', async () => {
    const response = await prism.process({
      method: 'get',
      url: {
        path: '/pet/findByStatus',
        query: {
          status: ['sold', 'available'],
        },
      },
    });
    expect(response.validations).toEqual({
      input: [],
      output: [],
    });
  });

  test('support param in body', async () => {
    const response = await prism.process({
      method: 'post',
      url: {
        path: '/store/order',
      },
      body: {
        id: 1,
        petId: 2,
        quantity: 3,
        shipDate: '12-01-2018',
        status: 'placed',
        complete: true,
      },
    });
    expect(response.validations).toEqual({
      input: [],
      output: [],
    });
  });

  test("should forward the request correctly even if resources haven't been provided", async () => {
    // Recreate Prism with no loaded document
    prism = createInstance(undefined, { forwarder, router: undefined, mocker: undefined });

    const response = await prism.process({
      method: 'post',
      url: {
        path: '/store/order',
        baseUrl: 'https://petstore.swagger.io',
      },
      body: {
        id: 1,
        petId: 2,
        quantity: 3,
        shipDate: '12-01-2018',
        status: 'placed',
        complete: true,
      },
    });

    expect(response.validations).toEqual({
      input: [],
      output: [],
    });
  });

  test('loads spec provided in yaml', async () => {
    prism = createInstance();
    await prism.load({
      path: relative(process.cwd(), resolve(__dirname, 'fixtures', 'petstore.oas2.yaml')),
    });

    expect(prism.resources).toHaveLength(3);
  });

  test('returns stringified static example when one defined in spec', async () => {
    prism = createInstance();
    await prism.load({
      path: relative(process.cwd(), resolve(__dirname, 'fixtures', 'static-examples.oas2.json')),
    });

    const response = await prism.process({
      method: 'get',
      url: {
        path: '/todos',
      },
    });

    expect(response.output).toBeDefined();
    expect(response.output!.body).toBeInstanceOf(Array);
  });
});
