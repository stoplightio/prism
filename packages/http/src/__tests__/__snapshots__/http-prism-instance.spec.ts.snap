// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Http Prism Instance function tests given correct route should return correct response 1`] = `
Object {
  "input": Object {
    "method": "get",
    "url": Object {
      "path": "/pet/findByStatus",
      "query": Object {
        "status": Array [
          "available",
          "pending",
        ],
      },
    },
  },
  "output": Object {
    "body": Anything,
    "headers": Object {
      "Content-type": "application/json",
    },
    "statusCode": 200,
  },
  "validations": Object {
    "input": Array [],
    "output": Array [],
  },
}
`;
